drivers/gles3/shaders/scene.glsl

```diff
@@ -971,6 +974,24 @@ void main() {
 #CODE : FRAGMENT
 	}
 
+// Normalize these only once if the user fragment shader set them
+// FIXME: It would be nice to have NORMAL_GET and NORMAL_SET instead to separate the read and write cases
+#ifdef NORMAL_USED
+	normal = normalize(normal);
+#endif
+#ifdef TANGENT_USED
+	tangent = normalize(tangent);
+	binormal = normalize(binormal);
+#endif
```

Comment:
We will need to run these shaders through radeon GPU analyzer (available on [shader playground](https://shader-playground.timjones.io/)) to see how this affects occupancy usage. Right now Godot is heavily bottlenecked by shader occupancy (i.e. we are using too many vector registers). Optimizations like this help reduce the number of instructions the shader does, but it may be at the cost of more VGPRs. We should check if it actually increases occupancy usage or not. (It may not if normal_interp was already taking a VGPR the shader compiler may be smart enough to drop that VGPR if normal_interp never gets accessed again)

Reply:
Good point, I'm not aware of these low level details yet.

How can you extract the substituted GLSL source code which is compiled?

Is there a way to save the generated GLSL code into the `.godot/shader_cache`, so it can be analyzed/checked?

Reply:
I don't think so. I have always just copy-pasted out of the source file and then added defines + custom lines of code as needed

Reply:
I'm running a master build, but cannot find any menu item to inspect/see the generated shader code. Where can I find it?

Reply:
Using this PR in practice: Godot 4 (left side) vs MagicaVoxel (right side, raytraced render)
![image](https://user-images.githubusercontent.com/1848514/193065707-779d54ac-bb82-4677-a8d1-761f449fcb6d.png)
Transparent and metallic are not prefect yet, but getting closer.

Reply:
Another example:
<img width="548" alt="MagicaVoxel-rendered-in-Godot4" src="https://user-images.githubusercontent.com/1848514/193114293-377de4c0-9afe-4670-835d-90db2af30d63.png">
Yes, this is a Godot 4 camera.

Reply:
> We will need to run these shaders through Radeon GPU analyzer (available on [shader playground](https://shader-playground.timjones.io/)) to see how this affects occupancy usage.

Godot yields 24 versions of the same shader. What are these versions and which one to run through the analyzer?

Reply:
The versions are just different combinations of preprocessor defines. You can just test with the combinations that are relevant to the changes you made.